### ISSUE TITLE
Feature Request: expose binary, library and configuration file paths as automatic attributes per-platform

### ISSUE THREAD

**Issue by bixu** [chef/ohai#444]
I have used a cookbook I've written to accomplish this in the past (https://github.com/bixu/paths), but I feel like this would be better provided by ohai.  Accessing these attributes might then look like this:

```
node.default['my_custom_binpath'] = node['binary_path'] + ":/some/string/path"

execute "build my_app" do
  command "make && make install"
  cwd '/path/to/my/src'
  environment => (
  'PATH' => node['my_custom_binpath'],
  'LD_LIBRARY_PATH' => node['library_path']
  )
  not_if "ls #{node['etc_path']}/my_app.conf"
end

```

Having this built into ohai would go a long way towards fulfilling the promise of truly cross-platform automation (80-90% of the issues I wrestle with on non-Ubuntu platforms are just pathing).

I'm very happy to make a PR with appropriate additions if the Chef folks/community think this is a good idea and that ohai is the right place to abstract this.

(CC'ing some folks I've mentioned this to in the past)

cc/ @sax @meatballhat @juliandunn

**Comment by lamont-granquist** [comment]
@someara would this reduce duplication and conditionals in the community cookbooks?

**Comment by someara** [comment]
I remember liking this idea last year when I first saw it, but it doesn't really have anything to do with the current approach I'm taking... not a huge fan of node attributes at the moment.

I'll have to take a closer look at this after I crank a few more cross-platform things out.

**Comment by lamont-granquist** [comment]
It might be better to put sugar methods into the Chef DSL for this?

**Comment by bixu** [comment]
@someara, I'd *love* to discuss this with you in more detail if you'd be up for that.  I'm on Berlin time (as I think you know), but can make myself available for a chat over IRC/Hangouts, whatever if it's helpful.

Generally excited to see this as it would move us a long ways towards the promise of one cookbook artifact for all platforms.

**Comment by lamont-granquist** [comment]
Yeah I don't see any value in having these and node attributes and there's too much that can go wrong.  For changing the defaults we should probably wire it up to Chef::Config so that users can override them if they need to (if users really desire to build a footgun then the chef-client cookbook can drop config settings based on node attributes and people can go to town).  And these should be DSL methods that come from core chef and not ohai.  I simply don't want to see the pain that'll come from people hacking on the default values with node attributes and hosing themselves with compile/converge problems or computable attributes issues.

**Comment by bixu** [comment]
I'm a little unclear how these new DSL methods might look - maybe some pseudo-code to show me what you are thinking?  Or do you mean simply exposing a method that would be used in a way similar to how we use Chef::Config[:file_cache_path]?

I do very much understand the desire to protect people from building footguns :) - that's in fact why I was suggesting this for Ohai.  I'm happy to submit a feature request against a different repo if you can specify which one you prefer.

**Comment by lamont-granquist** [comment]
like:

```
path_helper :bindir  # returns '/usr/bin'
path_helper :binpath # returns '/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin'
```

then you just define one method `def path_helper` in a DSL mixin.

its also even better to do this:

```
Chef.path_helper :bindir
```

then you're not overriding anyone's variables or methods that they've defined in their recipes named 'path_helper'

we don't currently have any examples of doing the latter, but its being discussed in an active RFC right now.

**Comment by bixu** [comment]
Very cool.  I look forward to using this!

**Comment by tas50** [comment]
Thanks for submitting this issue. After review by the team, we've decided that this isn't something that we're going to work on ourselves. We receive dozens of new issues a month, which is more than our team can possibly work on. This means we must prioritize work and be honest about what we're able to get done as a team.  That doesn't mean that we think this is a bad idea or isn't a necessary change. It's just not something we'll likely get to given competing priorities. I've applied our `Design Proposal: Accepted. PRs Welcome` label to this issue and closed it out. If you or someone else would like to work on this issue we're open to pull requests. If you'd like to chat further about what that design might look reach out to us in our [#chef-infra-dev](https://chefcommunity.slack.com/archives/C55FCUGBS) slack channel.
